_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,382 @@
+
+### The Prolog interpreter
+
+The first version of Erlang was implemented in Prolog in 1986. That
+version of Erlang was too slow for creating real applications, but it
+was useful for finding out which features of the languages were useful
+and which were not. New languages features could be added or deleted
+in a matter of hours or days.
+
+It soon became clear that Erlang needed to be at least 40 times faster
+to be useful in real projects.
+
+### JAM (Joe's Abstract Machine)
+
+In 1989 JAM (Joe's Abstract Machine) was first
+implemented. [Mike Williams][mike] wrote the runtime system
+in C, [Joe Armstrong][joe] wrote the compiler, and
+[Robert Virding][robert] wrote the libraries.
+
+[mike]: http://www.erlang-factory.com/conference/ErlangUserConference2013/speakers/MikeWilliams
+[joe]: https://github.com/joearms
+[robert]: https://github.com/rvirding
+
+JAM was 70 times faster than the Prolog interpreter, but it turned out
+that still wasn't fast enough.
+
+### TEAM (Turbo Erlang Abstract Machine)
+
+Bogumil ("Bogdan") Hausman created TEAM (Turbo Erlang Abstract
+Machine). It compiled the Erlang code to C code, which was then
+compiled to native code using GCC.
+
+It was significantly faster than JAM for small projects.
+Unfortunately, compilation was very slow, and the code size of the
+compiled code was too big to make it useful for large projects.
+
+### BEAM (Bogdan's Erlang Abstract Machine)
+
+Bogumil Hausman next machine was called BEAM (Bogdan's Erlang Abstract
+Machine). It was a hybrid machine that could execute both native code
+(translated via C) and [threaded code] with an [interpreter]. That
+allowed customers to compile their time-critical modules to native code
+and all other modules to threaded BEAM code. The threaded BEAM in
+itself was faster than JAM code.
+
+[threaded code]: https://en.wikipedia.org/wiki/Threaded_code
+[interpreter]: https://en.wikipedia.org/wiki/Interpreter_(computing)
+
+### Lessons Learned from BEAM/C
+
+The modern BEAM only has the interpreter. The ability of BEAM to generate
+C code was dropped in OTP R4. Why?
+
+C is not a suitable target language for an Erlang compiler. The main
+reason is that an Erlang function can't simply be translated to a C
+function because of Erlang's process model. Each Erlang process must
+have its own stack and that stack cannot be automatically managed by
+the C compiler.
+
+BEAM/C generated a single C function for each Erlang module. Local
+calls within the module were made by explicitly pushing the return
+address to the Erlang stack followed by a `goto` to the label of the
+called function.  (Strictly speaking, the calling function stores the
+return address to BEAM register and the called function pushes that
+register to the stack.)
+
+Calls to other modules were done similarly by using the GCC extension
+making it possible to take the address of a label and later jumping to
+it.  Thus an external call was made by pushing the return address to
```

Comment:
extension making it possible =>"GCC extension that makes it possible" ?

I think you can link to https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html so it is clear which extension you mean.